kong/init.lua

```diff
@@ -98,9 +99,95 @@ local ffi = require "ffi"
 local cast = ffi.cast
 local voidpp = ffi.typeof("void**")
 
+local plugins_map_version
+local plugins_map_semaphore
+
+local PLUGINS_MAP_CACHE_OPTS = { ttl = 0 }
+
+local configured_plugins
 local loaded_plugins
 local schema_state
 
+local function build_plugins_map(db, version)
+  local map = {}
+
+  for plugin, err in db.plugins:each() do
```

Comment:
I am concerned by this, in the sense that it can significantly increase the startup time for Kong. Routes and Services in Cassandra are partitioned together (i.e. they reside on the same node), but Plugins are not partitioned. This means that for every 100 plugins, we will make subsequent queries to the cluster to retrieve all rows from all partitions. On top of being potentially slow (we know of some users having in the (tens of) thousands of plugins), this also increases the chances of a failure to start due to a query timeout or such. We are most likely not in a position to experience any pain from this change, but I am fairly confident that some users will.

The impact of the query at runtime is less concerning I believe, as it will rarely be needed to rebuild the plugins map in actual production usage (and other operations esp. in the load balancer are making use of a complete iterator like this one as well).

I would suggest making a custom query for this operation (the new DAO is quite extensible and supports such custom operation per entity), and that would be helpful for PostgreSQL as well, **however**, I am really unsure there actually _is_ any improved query we could make for Cassandra... I am referring to something like:

```
SELECT DISTINCT name from plugins;
```

Unfortunately, our partitioning of the plugins column family isn't allowing us to do so in Cassandra... Ugh. I haven't found a workaround quite yet, or thought of an alternative approach.

What concerns do you have about the concern raised here?

Reply:
Yes, what you suggested for C* is basically what you usually do with proper NoSQL design, e.g. you have to start denormalize when you need to have read performance. But atm I do not see us doing that a lot as it introduces insert/update complexities.

About custom strategies, here is example:
https://github.com/Kong/kong/blob/master/kong/db/strategies/cassandra/routes.lua

And here is example of doing generic dao customization:
https://github.com/Kong/kong/blob/master/kong/db/dao/consumers.lua

Reply:
Just leaving a thought and not a review.

The concerns raised here are valid but if I'm not wrong the pattern already exists in the code.

While going through this thread, I recalled that when Kong starts up, we check if all plugins configured in DB are enabled or not, to guard against:
- A custom plugin which was enabled before but is not anymore
- With `plugins` property in Kong 0.14.0, someone could disable a bundled plugin after configuring it in Kong.

In both cases, Kong will fail to start up.

The code I'm referring to in master branch is [here](https://github.com/Kong/kong/blob/master/kong/init.lua#L110) and the next branch is [here](https://github.com/Kong/kong/blob/next/kong/db/dao/plugins.lua#L43). I see that it has undergone change with the plugins being moved to the new DAO.

Since both, this PR and the check plugins on startup logic need a set of plugin names that are configured in the DB, we could optimize this and load the plugins names from DB once (rather than in both places as happening currently), cache them and then use them to configure this PR and also verify if Kong should exit.

This is just a thought on how we can optimize the code and not an ask of any sorts.

Thoughts?

Reply:
Ugh, yes indeed, that is correct! I am not sure if it is a good thing or a bad thing, tbh, but at the very least we could take advantage of it, indeed!

Reply:
The iterator invoked in the `next` branch for the plugins load uses the same iterator in question here. A large part of the design of this PR was to use existing facilities in preparation for a stable release- if these facilities are not sufficient, we can address that. 

I would be much more concerned about optimizing the runtime query times as opposed to startup- the data access model in Kong already means that slow start concerns are the responsibility of infrastructure, and not software, architects (and I would agree). That given, the complexity and confusion to optimize the load-time query pattern is not worth it.

Here’s my semi final thought in this: if we are deeply concerned about performance in this context, we should build strategy-specific solutions, both of which had been alluded to here. Otherwise, we should acknowledge the utility and value of Kong’s underlying data access patterns, and leave it at that(and any potential optimization’s to that central code can come outside the scope of this pr).